google/auth/credentials.py

```diff
@@ -0,0 +1,209 @@
+    construction. Some classes will provide mechanisms to copy the credentials
+    with modifications such as :meth:`ScopedCredentials.with_scopes`.
+    """
+    def __init__(self):
+        self.token = None
+        """str: The bearer token that can be used in HTTP headers to make
+        authenticated requests."""
+        self.expiry = None
+        """Optional[datetime]: When the token expires and is no longer valid.
+        If this is None, the token is assumed to never expire."""
+
+    @property
+    def valid(self):
+        """Checks the validity of the credentials.
+
+        This is True if the credentials have a :attr:`token` and the token
+        is not :attr:`expired`.
+        """
+        return self.token is not None and not self.expired
+
+    @property
+    def expired(self):
+        """Checks if the credentials are expired.
+
+        Note that credentials can be invalid but not expired becaue Credentials
+        with :attr:`expiry` set to None is considered to never expire.
+        """
+        now = _helpers.utcnow()
+        if self.expiry is None or self.expiry > now:
+            return False
+        else:
+            return True
+
+    @abc.abstractmethod
+    def refresh(self, request):
+        """Refreshes the access token.
+
+        Args:
+            request (google.auth.transport.Request): A callable used to make
+                HTTP requests.
+
+        Raises:
+            google.auth.exceptions.RefreshError: If the credentials could
+                not be refreshed.
+        """
+        # pylint: disable=missing-raises-doc
+        # (pylint doesn't recognize that this is abstract)
+        raise NotImplementedError('Refresh must be implemented')
+
+    def apply(self, headers, token=None):
+        """Apply the token to the authentication header.
+
+        Args:
+            headers (Mapping): The HTTP request headers.
+            token (Optional[str]): If specified, overrides the current access
+                token.
+        """
+        headers[b'authorization'] = 'Bearer {}'.format(
+            _helpers.from_bytes(token or self.token))
+
+    def before_request(self, request, method, url, headers):
+        """Performs credential-specific before request logic.
+
+        Refreshes the credentials if necessary, then calls :meth:`apply` to
+        apply the token to the authentication header.
+
+        Args:
+            request (google.auth.transport.Request): A callable used to make
+                HTTP requests.
+            method (str): The request's HTTP method.
+            url (str): The request's URI.
+            headers (Mapping): The request's headers.
+        """
+        # pylint: disable=unused-argument
+        # (Subclasses may use these arguments to ascertain information about
+        # the http request.)
+        if not self.valid:
+            self.refresh(request)
+        self.apply(headers)
+
+
+@six.add_metaclass(abc.ABCMeta)
+class ScopedCredentials(object):
```

Comment:
> I'm not following the diamond inheritance problem worry. If all three interfaces contain no concrete elements (not even private ones), there's no diamond inheritance problem, right? It's not like what would be the two subinterfaces have conflicting method names.

Yes but the MRO would still be ... weird. I'd like to avoid it if possible.

> For common behavior, implement it once in one (private) place and delegate to it from the n places where it is exposed.

I could be on board with this but IMO it's separation without much purpose. The thing I really dislike about it how how the documentation will present the base classes:

With the current implementation:

<img width="681" alt="screen shot 2016-10-11 at 4 58 04 pm" src="https://cloud.githubusercontent.com/assets/250995/19292922/7cd9b232-8fd4-11e6-9347-ff49b58ca7ba.png">

With inheriting from a private concrete class:

<img width="683" alt="screen shot 2016-10-11 at 4 59 49 pm" src="https://cloud.githubusercontent.com/assets/250995/19292924/933fba26-8fd4-11e6-94f1-49552fab945d.png">

Any recommendations here? Failing that, is there any way you could be comfortable with keeping some of the "concreteness" around here? One thing I can do is mark *everything* as `abstractmethod` or `abstractproperty` and that forces all subclasses to either implement or explicitly call super.